core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.currentVersionId = base.currentVersionId();
+      this.versions = Lists.newArrayList(base.versions());
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder upgradeFormatVersion(int newFormatVersion) {
+      Preconditions.checkArgument(
+          newFormatVersion >= this.formatVersion,
+          "Cannot downgrade v%s view to v%s",
+          formatVersion,
+          newFormatVersion);
+
+      if (this.formatVersion == newFormatVersion) {
+        return this;
+      }
+
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      Preconditions.checkArgument(null != newLocation, "Invalid location: null");
+      if (null != this.location && this.location.equals(newLocation)) {
+        return this;
+      }
 
-    int versionHistorySizeToKeep =
-        PropertyUtil.propertyAsInt(
-            properties(),
-            ViewProperties.VERSION_HISTORY_SIZE,
-            ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
+      if (newVersionId == -1) {
+        ValidationException.check(
+            lastAddedVersionId != null,
+            "Cannot set last version id: no current version id has been set");
+        return setCurrentVersionId(lastAddedVersionId);
+      }
+
+      if (this.currentVersionId == newVersionId) {
+        return this;
+      }
+
+      checkCurrentVersionIdIsValid(newVersionId);
+
+      // by that time, schemas are defined and the highestFieldId can be determined and schema
+      // changes can be added
+      int highestFieldId = highestFieldId();
+      for (Schema schema : schemas) {
+        this.changes.add(new MetadataUpdate.AddSchema(schema, highestFieldId));
+      }
+
+      this.currentVersionId = newVersionId;
 
-    if (versionHistorySizeToKeep <= 0) {
-      LOG.warn(
-          "{} must be positive but was {}",
+      if (lastAddedVersionId != null && lastAddedVersionId == newVersionId) {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(LAST_ADDED));
+      } else {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      }
+
+      return this;
+    }
+
+    private int highestFieldId() {
+      return schemas.stream().map(Schema::highestFieldId).max(Integer::compareTo).orElse(0);
+    }
+
+    public Builder setCurrentVersion(ViewVersion version) {
+      Schema schema = indexSchemas(schemas).get(version.schemaId());
+      int newSchemaId = addSchemaInternal(schema);
```

Comment:
This doesn't seem correct to me.

The `ViewVersion` references a schema by ID. This method assumes that schema has been added to the builder already (or was already added to the view in a previous commit) because it expects the schema to be present in the ID to schema map produced by `indexSchemas`. But this then adds the schema and possibly reassigns the schema ID, which will be a noop because the schema is already known.

I think the solution is to pass the schema into this method. Then this would add the schema via `addSchemaInternal` (possibly reassigning its ID), add the view version that references the schema ID returned by `addSchemaInternal`, and then call `setCurrentVersionId` with the view version ID returned by `addVersionInternal`.